File: frontend/src/api/historicalAnalytics.ts
/**
 * API client functions for Historical Traffic Intelligence & Trend Analysis.
 * Phase 22: Historical Traffic Intelligence & Trend Analysis.
 */
import { apiClient } from '@/api/client';
import {
  AnomalyHistoryResponse,
  DirectionalTrendResponse,
  HistoricalFilterParams,
  HistoricalSummaryResponse,
  HistoricalTimeSeriesResponse,
  LaneIntelligenceResponse,
  PeakPeriodsResponse,
  PeriodComparisonResponse,
  SourceComparisonResponse,
  VehicleCompositionTrendResponse,
} from '@/types/historicalAnalytics';

function buildQueryString(params?: HistoricalFilterParams): string {
  if (!params) return '';
  const searchParams = new URLSearchParams();

  if (params.time_preset) searchParams.set('time_preset', params.time_preset);
  if (params.start_time) searchParams.set('start_time', params.start_time);
  if (params.end_time) searchParams.set('end_time', params.end_time);
  if (params.camera_source_id) searchParams.set('camera_source_id', params.camera_source_id);
  if (params.session_mode) searchParams.set('session_mode', params.session_mode);
  if (params.include_synthetic !== undefined) {
    searchParams.set('include_synthetic', String(params.include_synthetic));
  }
  if (params.bucket_interval) searchParams.set('bucket_interval', params.bucket_interval);

  const qs = searchParams.toString();
  return qs ? `?${qs}` : '';
}

/**
 * Fetches high-level historical traffic summary.
 */
export async function getHistoricalSummary(
  params?: HistoricalFilterParams
): Promise<HistoricalSummaryResponse> {
  const qs = buildQueryString(params);
  return apiClient<HistoricalSummaryResponse>(`/api/v1/historical-analytics/summary${qs}`);
}

/**
 * Fetches discrete non-interpolated time-series buckets.
 */
export async function getHistoricalTimeSeries(
  params?: HistoricalFilterParams
): Promise<HistoricalTimeSeriesResponse> {
  const qs = buildQueryString(params);
  return apiClient<HistoricalTimeSeriesResponse>(`/api/v1/historical-analytics/timeseries${qs}`);
}

/**
 * Fetches vehicle classification breakdown and trends over time.
 */
export async function getVehicleCompositionTrends(
  params?: HistoricalFilterParams
): Promise<VehicleCompositionTrendResponse> {
  const qs = buildQueryString(params);
  return apiClient<VehicleCompositionTrendResponse>(
    `/api/v1/historical-analytics/vehicle-composition${qs}`
  );
}

/**
 * Fetches directional traffic split and flow trends.
 */
export async function getDirectionalTrends(
  params?: HistoricalFilterParams
): Promise<DirectionalTrendResponse> {
  const qs = buildQueryString(params);
  return apiClient<DirectionalTrendResponse>(`/api/v1/historical-analytics/directions${qs}`);
}

/**
 * Fetches lane-level utilization, occupancy, and image-space density history.
 */
export async function getLaneIntelligence(
  params?: HistoricalFilterParams
): Promise<LaneIntelligenceResponse> {
  const qs = buildQueryString(params);
  return apiClient<LaneIntelligenceResponse>(`/api/v1/historical-analytics/lanes${qs}`);
}

/**
 * Fetches deterministic observed peak traffic periods.
 */
export async function getObservedPeakPeriods(
  params?: HistoricalFilterParams
): Promise<PeakPeriodsResponse> {
  const qs = buildQueryString(params);
  return apiClient<PeakPeriodsResponse>(`/api/v1/historical-analytics/peaks${qs}`);
}

/**
 * Fetches operational traffic anomaly and incident statistics.
 */
export async function getAnomalyHistory(
  params?: HistoricalFilterParams
): Promise<AnomalyHistoryResponse> {
  const qs = buildQueryString(params);
  return apiClient<AnomalyHistoryResponse>(`/api/v1/historical-analytics/anomalies${qs}`);
}

/**
 * Fetches comparative metrics grouped by camera or video source.
 */
export async function getSourceComparison(
  params?: HistoricalFilterParams
): Promise<SourceComparisonResponse> {
  const qs = buildQueryString(params);
  return apiClient<SourceComparisonResponse>(`/api/v1/historical-analytics/sources${qs}`);
}

/**
 * Fetches period-over-period delta comparisons.
 */
export async function getPeriodComparison(
  params?: HistoricalFilterParams
): Promise<PeriodComparisonResponse> {
  const qs = buildQueryString(params);
  return apiClient<PeriodComparisonResponse>(`/api/v1/historical-analytics/compare${qs}`);
}

File: backend/app/api/v1/historical_analytics.py
"""
Historical Traffic Intelligence & Trend Analysis API Router.
Phase 22: Historical Traffic Intelligence & Trend Analysis.

Mounted at: /api/v1/historical-analytics
Strictly descriptive, retrospective, provenance-grounded traffic intelligence.
"""
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.schemas.historical_analytics import (
    AnomalyHistoryResponse,
    DirectionalTrendResponse,
    HistoricalFilterParams,
    HistoricalSummaryResponse,
    HistoricalTimeSeriesResponse,
    LaneIntelligenceResponse,
    PeakPeriodsResponse,
    PeriodComparisonResponse,
    SourceComparisonResponse,
    VehicleCompositionTrendResponse,
)
from app.services.analytics.historical_analytics_service import HistoricalAnalyticsService

router = APIRouter()


def get_filter_params(
    start_time: Optional[datetime] = Query(None, description="Start datetime (ISO 8601 UTC)"),
    end_time: Optional[datetime] = Query(None, description="End datetime (ISO 8601 UTC)"),
    time_preset: Optional[str] = Query("7d", description="Preset: 24h, 7d, 30d, 90d, custom"),
    camera_source_id: Optional[str] = Query(None, description="Filter by camera source ID"),
    session_mode: Optional[str] = Query(None, description="Filter by mode: file_analysis, live_monitoring"),
    include_synthetic: bool = Query(False, description="Include synthetic/test data (default false for real-world isolation)"),
    bucket_interval: Optional[str] = Query("hourly", description="Bucketing interval: 15m, 30m, hourly, daily, weekly"),
) -> HistoricalFilterParams:
    return HistoricalFilterParams(
        start_time=start_time,
        end_time=end_time,
        time_preset=time_preset,
        camera_source_id=camera_source_id,
        session_mode=session_mode,
        include_synthetic=include_synthetic,
        bucket_interval=bucket_interval,
    )


@router.get(
    "/summary",
    response_model=HistoricalSummaryResponse,
    summary="Get aggregated historical traffic summary",
    description="Returns total observed volume, duration, normalized flow rates, inbound/outbound split, and provenance.",
)
def get_historical_summary(
    params: HistoricalFilterParams = Depends(get_filter_params),
    db: Session = Depends(get_db),
) -> HistoricalSummaryResponse:
    return HistoricalAnalyticsService.get_summary(db=db, params=params)


@router.get(
    "/timeseries",
    response_model=HistoricalTimeSeriesResponse,
    summary="Get historical discrete time-series buckets",
    description="Returns non-interpolated time-series bucket aggregation (hourly/daily) with volume, flow rates, and duration.",
)
def get_historical_timeseries(
    params: HistoricalFilterParams = Depends(get_filter_params),
    db: Session = Depends(get_db),
) -> HistoricalTimeSeriesResponse:
    return HistoricalAnalyticsService.get_timeseries(db=db, params=params)


@router.get(
    "/vehicle-composition",
    response_model=VehicleCompositionTrendResponse,
    summary="Get vehicle classification composition and trends",
    description="Returns volume, percentage share, and temporal trends strictly across genuine supported vehicle classes.",
)
def get_vehicle_composition_trends(
    params: HistoricalFilterParams = Depends(get_filter_params),
    db: Session = Depends(get_db),
) -> VehicleCompositionTrendResponse:
    return HistoricalAnalyticsService.get_vehicle_composition(db=db, params=params)


@router.get(
    "/directions",
    response_model=DirectionalTrendResponse,
    summary="Get directional traffic split and flow trends",
    description="Returns inbound vs. outbound traffic counts, percentage splits, and ratio trends over time.",
)
def get_directional_trends(
    params: HistoricalFilterParams = Depends(get_filter_params),
    db: Session = Depends(get_db),
) -> DirectionalTrendResponse:
    return HistoricalAnalyticsService.get_directions(db=db, params=params)


@router.get(
    "/lanes",
    response_model=LaneIntelligenceResponse,
    summary="Get lane utilization, occupancy, and density history",
    description="Returns historical lane volume share, peak/avg occupancy, and image-space density with calibration warnings.",
)
def get_lane_intelligence(
    params: HistoricalFilterParams = Depends(get_filter_params),
    db: Session = Depends(get_db),
) -> LaneIntelligenceResponse:
    return HistoricalAnalyticsService.get_lanes(db=db, params=params)


@router.get(
    "/peaks",
    response_model=PeakPeriodsResponse,
    summary="Get deterministic observed peak traffic periods",
    description="Identifies observed peak flow, peak volume, and peak density periods with explicit tie-breaking rules.",
)
def get_observed_peak_periods(
    params: HistoricalFilterParams = Depends(get_filter_params),
    db: Session = Depends(get_db),
) -> PeakPeriodsResponse:
    return HistoricalAnalyticsService.get_peaks(db=db, params=params)


@router.get(
    "/anomalies",
    response_model=AnomalyHistoryResponse,
    summary="Get historical traffic anomaly and incident statistics",
    description="Returns operational anomaly counts by type, severity, status, and associated observation sources.",
)
def get_anomaly_history(
    params: HistoricalFilterParams = Depends(get_filter_params),
    db: Session = Depends(get_db),
) -> AnomalyHistoryResponse:
    return HistoricalAnalyticsService.get_anomalies(db=db, params=params)


@router.get(
    "/sources",
    response_model=SourceComparisonResponse,
    summary="Get comparative analytics across camera and video sources",
    description="Compares traffic volume, observation duration, flow rates, and incidents across monitored sources.",
)
def get_source_comparison(
    params: HistoricalFilterParams = Depends(get_filter_params),
    db: Session = Depends(get_db),
) -> SourceComparisonResponse:
    return HistoricalAnalyticsService.get_sources(db=db, params=params)


@router.get(
    "/compare",
    response_model=PeriodComparisonResponse,
    summary="Get period-over-period comparative metrics",
    description="Compares traffic metrics between current time window and immediately preceding equivalent window.",
)
def get_period_comparison(
    params: HistoricalFilterParams = Depends(get_filter_params),
    db: Session = Depends(get_db),
) -> PeriodComparisonResponse:
    return HistoricalAnalyticsService.get_comparison(db=db, params=params)
